frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryPlaceVisitStorage.java

```diff
@@ -71,6 +71,7 @@ public boolean updateUserMarkOrAddPlaceVisit(
       if (placesMap == null) {
         Map<String, PlaceVisitModel> newPlaceMap = new ConcurrentHashMap<>();
         newPlaceMap.put(placeVisit.placeId(), placeVisit);
+        storage.put(placeVisit.tripId(), newPlaceMap);
```

Comment:
For some reason I thought we had switched this to `ConcurrentHashMap`.

Reply:
Looks like this is a concurrent hash map, and we're using it without map-level synchronization in `addPlaceVisit` above. That means this synchronized block is not safe against interleavings from other methods. (In particular, it looks like `ConcurrentHashMap` uses bucket-level synchronization, though that's an implementation detail and not part of the public interface.)

This will need to be updated to use `compute` or similar instead of a `synchronized` block.

I'm also not able to add a comment about it on the line above, but there's a weird catch/rethrow structure in `addPlaceVisit` that should be simplified to a direct throw. (The `try/catch` structure should be removed entirely.)

Reply:
I was doing a catch rethrow thing because within compute, only Runtime Exceptions can be thrown, so I couldn't directly throw a PlaceAlreadyExists exception within the compute function

Reply:
Hmm...that makes sense, but the intent is difficult to read. I would have probably minted a new runtime exception type for that and then caught only that and translated it appropriately. Don't bother doing that here, but please add a comment where you throw the runtime exception and explain why you're doing so.